internal/docker-client/docker.go

```diff
@@ -0,0 +1,166 @@
+}
+
+// CreateAssignment with the given dockerfile and script, returns a unique assignment id.
+func (d *DockerClient) CreateImage(ctx context.Context, imageName string, imageTar []byte) error {
+	buildOptions := types.ImageBuildOptions{
+		Dockerfile: "Dockerfile", // optional, is the default
+		Tags:       []string{fmt.Sprintf("%s:latest", imageName)},
+	}
+	img := bytes.NewReader(imageTar)
+	// build image
+	_, err := d.cli.ImageBuild(ctx, img, buildOptions)
+	if err != nil {
+		return errors.Wrap(err, graderd.ErrFailedToBuildImage.Error())
+	}
+
+	return nil
+}
+
+// ListTasks lists all the active tasks associated with the assignment id in the docker host.
+func (d *DockerClient) ListTasks(ctx context.Context) ([]*grader_task.Task, error) {
+	containers, err := d.cli.ContainerList(context.Background(), types.ContainerListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	tasks := d.mp.Enumerate()
+	taskList := make([]*grader_task.Task, 0)
+
+	for _, c := range containers {
+		for _, t := range tasks {
+			if t.ContainerID == c.ID {
+				status := graderd.ParseContainerState(c.State)
+				if status == grader_task.StatusStarted {
+					taskList = append(taskList, &t)
+				}
+			}
+		}
+	}
+	return taskList, nil
+}
+
+//func (d *DockerClient) CreateTasks(ctx context.Context, taskList []*grader_task.Task, db dockerdb.Database) error {
+//	// create tasks
+//	for _, t := range taskList {
+//		err := d.createTask(ctx, t)
+//		if err != nil {
+//			return err
+//		}
+//	}
+//	// update database
+//	if err := db.PutTasks(ctx, taskList); err != nil {
+//		return errors.Wrap(err, ErrFailedToUpdateTask.Error())
+//	}
+//
+//	return nil
+//}
+//
+//func (d *DockerClient) createTask(ctx context.Context, task *grader_task.Task) error {
+//	// create container
+//	resp, err := d.cli.ContainerCreate(ctx, &container.Config{
+//		Image: task.ImageID,
+//		Labels: map[string]string{
+//			"student_name": task.StudentName,
+//		},
+//		StopTimeout: task.Timeout,
+//	}, nil, nil, task.ID)
+//	if err != nil {
+//		return errors.Wrap(err, ErrFailedToCreateTask.Error())
+//	}
+//
+//	// assign id and time
+//	task.ContainerID = resp.ID
+//	t := time.Now()
+//	task.CreatedTime = &t
+//	task.Status = grader_task.StatusPending
+//
+//	return nil
+//}
+
+// StartTasks starts execution of all the given tasks (container IDs).
+//func (d *DockerClient) StartTasks(ctx context.Context, taskList []*grader_task.Task, db dockerdb.Database) error {
+//	for _, task := range taskList {
+//		if err := d.cli.ContainerStart(ctx, task.ContainerID, types.ContainerStartOptions{}); err != nil {
+//			return errors.Wrap(err, ErrFailedToStartTask.Error())
+//		}
+//		// update dockerdb
+//		task.Status = grader_task.StatusStarted
+//		if err := db.UpdateTask(ctx, task); err != nil {
+//			return errors.Wrap(err, ErrFailedToUpdateTask.Error())
+//		}
+//	}
+//	return nil
+//}
+
+func (d *DockerClient) StartTask(ctx context.Context, taskID string) error {
+	return nil
```

Comment:
the important part is missing!!!

Reply:
I'm hammering that one out after I write the rest of it!